packages/ra-ui-materialui/src/button/SaveButton.tsx

```diff
@@ -101,7 +112,8 @@ export const SaveButton = <RecordType extends RaRecord = any>(
                 event.preventDefault();
                 const isFormValid = await form.trigger();
                 if (!isFormValid) {
-                    event.preventDefault();
+                    event.stopPropagation();
+                    await form.handleSubmit(handleSubmitWithErrors)(event);
```

Comment:
Just realizing we have done exactly the same fix @WiXSL , didn't know you were working on #7555  too 😅 

Anyway, I did it almost the same way as you, albeit with one difference: actually I use the `handleSubmit` callback both for the error case and the case when the form is valid.
This removes the need to duplicate code like `notify('ra.message.invalid_form', { type: 'warning' });` or `saveContext?.save(values, {...})`

Below is my implementation:

```tsx
    const handleSubmit = useCallback(
        async values => {
            console.log('here1');
            let errors;
            if (saveContext?.save) {
                console.log('here2');
                errors = await saveContext?.save(values, {
                    ...mutationOptions,
                    transform,
                });
            }
            if (errors != null) {
                console.log('here3');
                setSubmissionErrors(errors, form.setError);
            }
        },
        [form, saveContext, mutationOptions, transform]
    );
```

```tsx
    const handleClick: MouseEventHandler<HTMLButtonElement> = useCallback(
        async event => {
            if (onClick) {
                onClick(event);
            }
            if (event.defaultPrevented) {
                return;
            }
            if (type === 'button') {
                // Prevent outer forms to receive the event
                event.stopPropagation();
                form.handleSubmit(handleSubmit)(event);
                return;
            }
        },
        [onClick, type, form, handleSubmit]
    );
```

What do you think?

Reply:
Yes, actually, I was going to improve it like you did it, without calling `form.trigger()`